plutus-tx/src/PlutusTx/Blueprint/Definition.hs

```diff
@@ -0,0 +1,78 @@
+{-# LANGUAGE AllowAmbiguousTypes      #-}
+{-# LANGUAGE DataKinds                #-}
+{-# LANGUAGE FlexibleContexts         #-}
+{-# LANGUAGE FlexibleInstances        #-}
+{-# LANGUAGE MultiParamTypeClasses    #-}
+{-# LANGUAGE PolyKinds                #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
+{-# LANGUAGE TypeApplications         #-}
+{-# LANGUAGE TypeFamilies             #-}
+{-# LANGUAGE TypeOperators            #-}
+{-# LANGUAGE UndecidableInstances     #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+{-# OPTIONS_GHC -Wno-unticked-promoted-constructors #-}
+
+-- | This module provides a functionality to derive and reference schema definitions.
+module PlutusTx.Blueprint.Definition (
+  module DefinitionId,
+  HasSchemaDefinition,
+  definitionRef,
+  deriveSchemaDefinitions,
+) where
+
+import Data.Kind (Constraint, Type)
+import Data.Map (Map)
+import Data.Map qualified as Map
+import GHC.TypeLits qualified as GHC
+import PlutusTx.Blueprint.Class (HasSchema, schema)
+import PlutusTx.Blueprint.Definition.Id as DefinitionId
+import PlutusTx.Blueprint.Schema (Schema (..))
+
+-- | Construct a schema that is a reference to a schema definition.
+definitionRef ::
+  forall typ types.
+  (AsDefinitionId typ, HasSchemaDefinition typ types) =>
+  Schema types
+definitionRef = SchemaDefinitionRef (definitionId @typ)
+
+{- |
+  A constraint that checks if a schema definition is present in a list of schema definitions.
+  Gives a user-friendly error message if the schema definition is not found.
+-}
+type HasSchemaDefinition :: Type -> k -> Constraint
+type family HasSchemaDefinition n xs where
+  HasSchemaDefinition x (x ': xs) = ()
+  HasSchemaDefinition x (_ ': xs) = HasSchemaDefinition x xs
+  HasSchemaDefinition n xs =
+    GHC.TypeError
+      ( GHC.ShowType n
+          GHC.:<>: GHC.Text " type was not found in the list of types having schema definitions."
+      )
+
+-- | Derive a map of schema definitions from a list of types.
+deriveSchemaDefinitions ::
+  forall (types :: [Type]).
+  (AsDefinitionsEntries types types) =>
+  Map DefinitionId (Schema types)
+deriveSchemaDefinitions = Map.fromList (definitionEntries @types @types)
+
+{- |
+  A class of types that can be converted to a list of schema definition entries.
+  It is used internally to derive a map of schema definitions from a list of types.
+-}
+class AsDefinitionsEntries (allTypes :: [Type]) (remainingTypes :: [Type]) where
```

Comment:
What does `remainingTypes` mean? "remaining" in what sense?

Reply:
Maybe you can suggest a better name, the idea is the following:

2 instances (one for a recursive step and one for a termination) iterate a type level list, taking a head and tail at each step,  checking if the head is in the `allTypes` and recursing on tail `remainingTypes`. This way in the beginning of iteration `allTypes` == `remainingTypes` and then `allTypes` stays the same list, while `remainingTypes` is shrinking until empty.

Here is an analogy at the value level, where `remainingTypes` serves a similar purpose:
```haskell

type Typ = String
type DefinitionId = String
type Schema = String

asDefinitionEntries :: [Typ] -> [(DefinitionId, Schema)]
asDefinitionEntries allTypes = go allTypes allTypes
  where
    go :: [Typ] -> [Typ] -> [(DefinitionId, Schema)]
    go allTypes remainingTypes = 
      case remainingTypes of
        [] -> []
        (t : tail) -> 
          let defId = lookupDefinitionId t allTypes
              schema = lookupSchema t allTypes
          in (defId, schema) : go allTypes tail

lookupDefinitionId :: Typ -> [Typ] -> DefinitionId
lookupDefinitionId t allTypes | t `elem` allTypes = "DefinitionId for " ++ t
lookupDefinitionId t _ = error $ "Type " ++ show t ++ " not found"

lookupSchema :: Typ -> [Typ] -> Schema
lookupSchema t allTypes | t `elem` allTypes = "Schema for " ++ t
lookupSchema t _ = error $ "Type " ++ show t ++ " not found"
```